Sources/Sentry/SentryDefaultRateLimits.m

```diff
@@ -1,106 +1,95 @@
- */
-- (NSDate *)defaultRadioSilenceDeadline {
-    return [[SentryCurrentDate date] dateByAddingTimeInterval:60];
-}
-
-/**
- * parses value of HTTP Header "Retry-After" which in most cases is sent in
- * combination with HTTP status 429 Too Many Requests.
- *
- * Retry-After value is a time-delta in seconds or a date.
- * In every case this method computes the date aka. `radioSilenceDeadline`.
- *
- * See RFC2616 for details on "Retry-After".
- * https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.37
- *
- * @return NSDate representation of Retry-After.
- *         As fallback `defaultRadioSilenceDeadline` is returned if parsing was
- *         unsuccessful.
- */
-- (NSDate *)parseRetryAfterHeader:(NSString * __nullable)retryAfterHeader {
-    if (nil == retryAfterHeader || 0 == [retryAfterHeader length]) {
-        return [self defaultRadioSilenceDeadline];
+- (BOOL)isRetryAfterHeaderLimitActive {
+    if (nil == self.retryAfterHeaderDate) {
+        return NO;
     }
-
-    NSDate *now = [SentryCurrentDate date];
-
-    // try to parse as double/seconds
-    double retryAfterSeconds = [retryAfterHeader doubleValue];
-    NSLog(@"parseRetryAfterHeader string '%@' to double: %f", retryAfterHeader, retryAfterSeconds);
-    if (0 != retryAfterSeconds) {
-        return [now dateByAddingTimeInterval:retryAfterSeconds];
+    
+    BOOL isActive = [self isInFuture:self.retryAfterHeaderDate];
+    if (isActive) {
+        [SentryLog logWithMessage:[NSString stringWithFormat:@"Retry-After limit reached until: %@",  self.retryAfterHeaderDate] andLevel:kSentryLogLevelDebug];
+        return YES;
+    } else {
+        return NO;
     }
+}
 
-    // parsing as double/seconds failed, try to parse as date
-    NSDate *retryAfterDate = [self.httpDateParser dateFromString:retryAfterHeader];
+- (BOOL)isInFuture:(NSDate *)date {
+    NSComparisonResult result = [[SentryCurrentDate date] compare:date];
+    if (result == NSOrderedAscending) {
+        return YES;
+    }
+    
+    return NO;
+}
 
-    if (nil == retryAfterDate) {
-        // parsing as seconds and date failed
-        return [self defaultRadioSilenceDeadline];
+- (void)update:(NSHTTPURLResponse *)response {
+    NSString *rateLimitsHeader = response.allHeaderFields[@"X-Sentry-Rate-Limits"];
+    if (nil != rateLimitsHeader) {
+        NSDictionary<NSString *, NSDate *> * limits = [self.rateLimitParser parse:rateLimitsHeader];
+        
+        @synchronized (self) {
+            [self.rateLimits addEntriesFromDictionary:limits];
+        }
+    } else if (response.statusCode == 429) {
+        NSDate* retryAfterHeaderDate = [self.retryAfterHeaderParser parse:response.allHeaderFields[@"Retry-After"]];
+        
+        if (nil == retryAfterHeaderDate) {
+            // parsing failed use default value
+            retryAfterHeaderDate = [[SentryCurrentDate date] dateByAddingTimeInterval:60];
+        }
+        
+        @synchronized (self) {
```

Comment:
We are setting this value under a lock but we're reading it outside of it.
I believe we either need to synchronize all access to this field (reader-writer lock would be best) or, if possible, just not synchronize at all. Now my knowledge of Obj-C fails here but is it correct to say:
NSDate being heap allocated and `retryAfterHeaderDate` being a reference to it. Are reference assignments guaranteed to be atomic? If so, we could get away with not synchronizing at all. If we need to perform two operations on the same value, we just copy the reference to the local scope:

```objective-c
NSDate *currentRetryAfterHeaderDate = self.retryAfterHeaderDate;
if (currentRetryAfterHeaderDate > x) {
// more access to currentRetryAfterHeaderDate over a consistent value
}
```

Simpler solution is just to throw `synchronize` everywhere and not bother

Reply:
All accesses to `rateLimits` and `retryAfterHeaderDate` are in sync blocks. Keep in mind that `isCustomHeaderRateLimitActive` and `isRetryAfterHeaderLimitActive` are private.